File: Map/web/src/lib/scoring.ts
// pipeline/scoring/scoring.py 와 계약 테스트로 동기화되는 스코어링 코어.
// 합산 순서까지 Python 구현과 동일하게 유지할 것 (부동소수 일치 1e-9).

export interface IndicatorRef {
  id: string;
  axis: string;
}

export interface RegionScoreInput {
  pct: Record<string, number | null>;
  flags: { metro: boolean; jeju: boolean };
}

export interface ScoreResult {
  axis: Record<string, number>;
  base: number | null;
  penalty: number;
  total: number | null;
  grade: string | null;
  coverage: number;
}

export const PENALTY_METRO = -15.0;
export const PENALTY_JEJU = -10.0;

const GRADES: Array<[number, string]> = [
  [85.0, "S"],
  [75.0, "A"],
  [65.0, "B"],
  [50.0, "C"],
];

export function gradeOf(total: number): string {
  for (const [cut, g] of GRADES) {
    if (total >= cut) return g;
  }
  return "D";
}

export function axisScores(region: RegionScoreInput, indicators: IndicatorRef[]): Record<string, number> {
  const byAxis: Record<string, number[]> = {};
  for (const ind of indicators) {
    const p = region.pct[ind.id];
    if (p === null || p === undefined) continue;
    (byAxis[ind.axis] ??= []).push(p);
  }
  const out: Record<string, number> = {};
  for (const a of Object.keys(byAxis)) {
    const v = byAxis[a];
    out[a] = v.reduce((s, x) => s + x, 0) / v.length;
  }
  return out;
}

export function computeScore(
  region: RegionScoreInput,
  indicators: IndicatorRef[],
  weights: Record<string, number>,
): ScoreResult {
  const ax = axisScores(region, indicators);
  const totalW = Object.values(weights).reduce((s, x) => s + x, 0);
  let availW = 0;
  for (const a of Object.keys(weights)) {
    if (a in ax) availW += weights[a];
  }
  if (totalW <= 0 || availW <= 0) {
    return { axis: ax, base: null, penalty: 0, total: null, grade: null, coverage: 0 };
  }
  let acc = 0;
  for (const a of Object.keys(ax)) {
    if (a in weights) acc += weights[a] * ax[a];
  }
  const base = acc / availW;
  let penalty = 0;
  if (region.flags?.metro) penalty += PENALTY_METRO;
  if (region.flags?.jeju) penalty += PENALTY_JEJU;
  const total = Math.min(100, Math.max(0, base + penalty));
  return { axis: ax, base, penalty, total, grade: gradeOf(total), coverage: availW / totalW };
}

export interface RankedRegion {
  code: string;
  total: number | null;
  rank: number | null;
}

export function rankAll(
  regions: Array<{ code: string } & RegionScoreInput>,
  indicators: IndicatorRef[],
  weights: Record<string, number>,
): Map<string, ScoreResult & { rank: number | null }> {
  const results = regions.map((r) => ({ code: r.code, res: computeScore(r, indicators, weights) }));
  const sorted = results
    .filter((x) => x.res.total !== null)
    .sort((a, b) => (b.res.total! - a.res.total!) || a.code.localeCompare(b.code));
  const rankByCode = new Map<string, number>();
  sorted.forEach((x, i) => rankByCode.set(x.code, i + 1));
  const out = new Map<string, ScoreResult & { rank: number | null }>();
  for (const x of results) out.set(x.code, { ...x.res, rank: rankByCode.get(x.code) ?? null });
  return out;
}

// 가중치 ±20% 섭동 → 순위 IQR → 안정성 1~5 (mulberry32 고정 시드로 재현 가능)
function mulberry32(seed: number) {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export function stabilityAll(
  regions: Array<{ code: string } & RegionScoreInput>,
  indicators: IndicatorRef[],
  weights: Record<string, number>,
  samples = 60,
  seed = 42,
): Map<string, { iqr: number; dots: number }> {
  const rng = mulberry32(seed);
  const axes = Object.keys(weights);
  const ranks = new Map<string, number[]>(regions.map((r) => [r.code, []]));
  for (let s = 0; s < samples; s++) {
    const w: Record<string, number> = {};
    for (const a of axes) w[a] = weights[a] * (0.8 + 0.4 * rng());
    const rr = rankAll(regions, indicators, w);
    for (const r of regions) {
      const rank = rr.get(r.code)?.rank;
      if (rank !== null && rank !== undefined) ranks.get(r.code)!.push(rank);
    }
  }
  const out = new Map<string, { iqr: number; dots: number }>();
  for (const [code, arr] of ranks) {
    if (arr.length === 0) {
      out.set(code, { iqr: Infinity, dots: 0 });
      continue;
    }
    const sorted = [...arr].sort((a, b) => a - b);
    const q = (p: number) => sorted[Math.min(sorted.length - 1, Math.floor(p * sorted.length))];
    const iqr = q(0.75) - q(0.25);
    const dots = iqr <= 2 ? 5 : iqr <= 5 ? 4 : iqr <= 12 ? 3 : iqr <= 25 ? 2 : 1;
    out.set(code, { iqr, dots });
  }
  return out;
}

File: Map/pipeline/scoring/scoring.py
"""스코어링 코어 — web/src/lib/scoring.ts 와 계약 테스트로 동기화되는 순수 함수들.

입력 규약:
  region = { "pct": {indicator_id: 0~100 | None}, "flags": {"metro": bool, "jeju": bool} }
  indicators = [{ "id": str, "axis": str }, ...]
  weights = {axis_id: number}  # 합이 1이 아니어도 됨(내부 정규화)

규칙:
  - 축 점수 = 축 내 존재하는 지표 pct 의 단순평균 (MVP: 축내 지표 등가중)
  - 총점 base = 존재하는 축들의 가중평균 (결측 축 가중치는 재분배 효과)
  - coverage = 존재 축 가중치 합 / 전체 가중치 합
  - 페널티: metro -15, jeju -10 → clamp(0, 100)
"""

PENALTY_METRO = -15.0
PENALTY_JEJU = -10.0

GRADES = [(85.0, "S"), (75.0, "A"), (65.0, "B"), (50.0, "C")]


def axis_scores(region, indicators):
    by_axis = {}
    for ind in indicators:
        p = region["pct"].get(ind["id"])
        if p is None:
            continue
        by_axis.setdefault(ind["axis"], []).append(p)
    return {a: sum(v) / len(v) for a, v in by_axis.items()}


def grade(total):
    for cut, g in GRADES:
        if total >= cut:
            return g
    return "D"


def compute_score(region, indicators, weights):
    ax = axis_scores(region, indicators)
    total_w = sum(weights.values())
    avail_w = sum(w for a, w in weights.items() if a in ax)
    if total_w <= 0 or avail_w <= 0:
        return {"axis": ax, "base": None, "penalty": 0.0, "total": None, "grade": None, "coverage": 0.0}
    base = sum(weights[a] * ax[a] for a in ax if a in weights) / avail_w
    penalty = 0.0
    if region.get("flags", {}).get("metro"):
        penalty += PENALTY_METRO
    if region.get("flags", {}).get("jeju"):
        penalty += PENALTY_JEJU
    total = min(100.0, max(0.0, base + penalty))
    return {
        "axis": ax,
        "base": base,
        "penalty": penalty,
        "total": total,
        "grade": grade(total),
        "coverage": avail_w / total_w,
    }


def percentile_scores(values, direction):
    """값 리스트(None 포함) → 0~100 백분위 점수. direction: 'up' 높을수록 좋음 / 'down' 낮을수록 좋음.

    평균 순위(mid-rank) 방식, n=1이면 50. None은 None 유지.
    """
    idx_vals = [(i, v) for i, v in enumerate(values) if v is not None]
    out = [None] * len(values)
    n = len(idx_vals)
    if n == 0:
        return out
    if n == 1:
        out[idx_vals[0][0]] = 50.0
        return out
    sorted_vals = sorted(v for _, v in idx_vals)
    # mid-rank: 동값 구간의 평균 순위
    first = {}
    count = {}
    for rank, v in enumerate(sorted_vals):
        if v not in first:
            first[v] = rank
            count[v] = 0
        count[v] += 1
    for i, v in idx_vals:
        mid = first[v] + (count[v] - 1) / 2.0
        p = mid / (n - 1) * 100.0
        out[i] = round(100.0 - p if direction == "down" else p, 2)
    return out
